include/transactions.hpp

```diff
@@ -0,0 +1,443 @@
+#ifndef GRAPH_PROTOTYPE_TRANSACTIONS_HPP
+#define GRAPH_PROTOTYPE_TRANSACTIONS_HPP
+
+#include "tag.hpp"
+#include <pmtv/pmt.hpp>
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <concepts>
+#include <functional>
+#include <list>
+#include <span>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <settings.hpp>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+#include <fmt/chrono.h>
+#pragma GCC diagnostic pop
+
+#include "utils.hpp"
+
+namespace fair::graph {
+
+static auto nullMatchPred = [](auto, auto) { return false; };
+
+template<typename Node>
+class ctx_settings : public settings_base {
+    using MatchPredicate                                = std::function<bool(const property_map &, const property_map &)>;
```

Comment:
As discussed during the meeting today: the existing OpenCMW implementation also supported partial matching, i.e. in case the perfect match hasn't been found the next best one in the (user-defined) hierarchy shall be returned, as illustrated, for example, [here](https://github.com/fair-acc/opencmw-cpp/pull/156). Specifically:

```cpp
opencmw::CtxSetting<int, std::string, 16, std::chrono::seconds, 3600 * 24, 10> settings; // history length of 24 hours, time-out of 10 seconds

assert(settings.commit(TimingCtx(1, 1, 1, 1), 101)); // store 101 for 'TimingCtx(cid=1, sid=1, pid=1, gid=1)'
assert(settings.commit(TimingCtx(1, 1, 1), 102));    // store 102 for 'TimingCtx(cid=1, sid=1, pid=1, gid=-1)'
assert(settings.commit(TimingCtx(1, 1), 103));       // store 103 for 'TimingCtx(cid=1, sid=1, pid=-1, gid=-1)'
assert(settings.commit(TimingCtx(1), 104));          // store 104 for 'TimingCtx(cid=1, sid=-1, pid=-1, gid=-1)'

// retrieve settings
assert(settings.get(TimingCtx(1)) == 104);
assert(settings.get(TimingCtx(1, 1, 1, 1)) == 101);
assert(settings.get(TimingCtx(1, 1, 1, 1)).timingCtx == TimingCtx(1, 1, 1, 1));
assert(settings.get(TimingCtx(1, 1, 1, 1)) == TimingCtx(1, 1, 1, 1)); // short-hand notation
assert(settings.get(TimingCtx(1, 1, 1, 2)) == 102); // no setting for 'gid=2' -> fall back to 'gid=-1'
assert(settings.get(TimingCtx(1, 1, 1, 2)) == TimingCtx(1, 1, 1));
assert(settings.get(TimingCtx(1, 1, 2, 2)) == 103); // no setting for 'pid=2' and 'gid=2' -> fall back to 'pid=gid=-1'
assert(settings.get(TimingCtx(1, 1, 2, 2)) == TimingCtx(1, 1));

// stage new settings
auto [committed, commitTimeStamp] = settings.commit(TimingCtx(1), 55); // immediately store 55 for 'TimingCtx(cid=1)'
assert(not settings.stage(TimingCtx(1), 555, "tokenNo1"));
assert(not settings.stage(TimingCtx(1), 556, "tokenNo2"));

assert(settings.get(TimingCtx(1)) == 55); // staged settings not yet commited
assert(settings.get(TimingCtx(2)) != TimingCtx(2));
assert(settings.commit("tokenNo1"));
assert(settings.get(TimingCtx(1)) == 555);
assert(settings.commit("tokenNo2"));
assert(settings.get(TimingCtx(1)) == 556);
```

If you could also add this as corresponding unit tests. 

As already discussed the simple true/false predicate might not be enough to implement this fall-back logic, maybe we could use a signature like

```cpp
  std::optional<bool>(const property_map &, const property_map &, std::size_t attempt = 0);
  // or complementary
 std::optional<bool>(const property_map &, const property_map &, std::size_t attempt = 0);
```
Where the attempt indicates the search iteration, e.g. 
 * `attempt==0` -> first try, perform a full match of e.g. `(bpcid, cid,sid, pid)`
 * `attempt==1` -> second try, perform a partial match of e.g. `(bpcid, cid,sid)`
 * `attempt==2` -> third try, perform a partial match of e.g. `(bpcid, cid)`
 * `attempt==3` -> fourth try, perform a partial match of e.g. `(bpcid)`
 * `attempt==4` -> fifth try, give up and return `std::nullopt`
 
Hope this is not to hackish but this should at least fit the FAIR specific use-case and hopefully be generic enough to extend to other applications. One could also query the test function for all cases to have only a single pass through all test cases and classify them according to a 'match score', e.g. '0<->perfect match`, '1<->one aspect did not match`. The latter would have the issue of not being explicit which hierarchical search parameter did not match... :thinking: 

Your feedback/improvement proposals would be welcome.

Many thanks in advance.